File: python_backend/report.py
"""会话 HTML 报告生成。"""
from __future__ import annotations

import pathlib
import time
from typing import Any

from session_data import SessionData

_CONC_DIR = pathlib.Path(__file__).parent.parent / "conc"


def _escape(text: str) -> str:
    return (
        text.replace("&", "&amp;")
        .replace("<", "&lt;")
        .replace(">", "&gt;")
        .replace('"', "&quot;")
        .replace("\n", "<br>")
    )


def generate_report(data: SessionData) -> pathlib.Path:
    duration_fmt = f"{data.duration:.1f}s"

    # Chat
    chat_parts = []
    for role, content in data.messages:
        alias = {"user": "用户", "assistant": "AI", "tool_card": "工具"}
        label = alias.get(role, role)
        chat_parts.append(
            f'<div class="msg {role}"><div class="label">{_escape(label)}</div>'
            f'<div class="bubble">{_escape(content)}</div></div>'
        )
    chat_html = "\n".join(chat_parts) or "<p>无聊天记录</p>"

    # Commands
    cmd_rows = []
    for ts, name, args, result in data.commands:
        tstr = time.strftime("%H:%M:%S", time.localtime(ts))
        arg_str = ", ".join(f"{k}={v}" for k, v in args.items())
        cmd_rows.append(
            f"<tr><td>{tstr}</td><td>{_escape(name)}</td>"
            f"<td>{_escape(arg_str)}</td><td>{_escape(result)}</td></tr>"
        )
    cmd_html = (
        "<table><tr><th>时间</th><th>指令</th><th>参数</th><th>结果</th></tr>"
        + "\n".join(cmd_rows)
        + "</table>"
    ) if cmd_rows else "<p>无控制指令</p>"

    html = f"""<!DOCTYPE html>
<html lang="zh-CN">
<head>
<meta charset="utf-8" />
<title>电机控制会话报告</title>
<style>
body {{ font-family: 'Noto Sans SC', system-ui, sans-serif; max-width: 800px; margin: 20px auto; padding: 0 20px; background: #f8f9fa; }}
h1, h2 {{ color: #333; border-bottom: 2px solid #dee2e6; padding-bottom: 6px; }}
table {{ border-collapse: collapse; width: 100%; margin: 12px 0; }}
th, td {{ border: 1px solid #dee2e6; padding: 8px 12px; text-align: left; }}
th {{ background: #e9ecef; }}
.stats {{ display: flex; gap: 16px; flex-wrap: wrap; }}
.stat-card {{ background: white; border: 1px solid #dee2e6; border-radius: 8px; padding: 16px; flex: 1; min-width: 120px; text-align: center; }}
.stat-card .value {{ font-size: 24px; font-weight: bold; color: #007bff; }}
.stat-card .label {{ font-size: 13px; color: #888; }}
.msg {{ margin: 8px 0; }}
.msg .label {{ font-size: 12px; color: #888; }}
.msg .bubble {{ display: inline-block; padding: 8px 14px; border-radius: 8px; max-width: 80%; line-height: 1.5; }}
.msg.user .bubble {{ background: #d1e7ff; }}
.msg.assistant .bubble {{ background: #ffffff; border: 1px solid #dee2e6; }}
.msg.tool_card .bubble {{ background: #fff3cd; border: 1px solid #ffc107; }}
</style>
</head>
<body>
<h1>📊 电机控制会话报告</h1>
<p>生成时间: {time.strftime("%Y-%m-%d %H:%M:%S")}</p>

<h2>运行统计</h2>
<div class="stats">
<div class="stat-card"><div class="value">{duration_fmt}</div><div class="label">运行时长</div></div>
<div class="stat-card"><div class="value">{data.max_speed:.0f}</div><div class="label">最高转速 (RPM)</div></div>
<div class="stat-card"><div class="value">{data.avg_speed:.0f}</div><div class="label">平均转速 (RPM)</div></div>
<div class="stat-card"><div class="value">{data.max_current:.2f}</div><div class="label">最大电流 (A)</div></div>
</div>

<h2>控制指令时间轴</h2>
{cmd_html}

<h2>完整聊天记录</h2>
{chat_html}
</body>
</html>"""

    _CONC_DIR.mkdir(parents=True, exist_ok=True)
    timestamp = time.strftime("%Y%m%d_%H%M%S")
    out_path = _CONC_DIR / f"session_report_{timestamp}.html"
    out_path.write_text(html, encoding="utf-8")
    return out_path

File: lib/report.ts
/**
 * 会话报告导出(浏览器版)— 替代 Electron 版 reportGenerator.ts(文件系统 + shell.open)。
 * 前端直接生成 HTML 并通过 Blob 下载,无需服务端。
 */
import type { Session } from './types'

function esc(s: string): string {
  return s.replace(/&/g, '&amp;').replace(/</g, '&lt;').replace(/>/g, '&gt;').replace(/"/g, '&quot;')
}

function buildHtml(session: Session): string {
  const rows = session.messages
    .map((m) => {
      const role = m.role === 'user' ? '用户' : m.role === 'assistant' ? 'AI' : '系统'
      const time = new Date(m.ts).toLocaleString('zh-CN')
      return `<div class="msg ${m.role}"><span class="tag">${role}</span><span class="time">${esc(time)}</span><div class="content">${esc(m.content)}</div></div>`
    })
    .join('')

  const stats = {
    messages: session.messages.length,
    user: session.messages.filter((m) => m.role === 'user').length,
    duration: Math.max(0, session.updatedAt - session.createdAt)
  }

  return `<!DOCTYPE html>
<html lang="zh-CN">
<head>
<meta charset="utf-8">
<title>会话报告 - ${esc(session.title)}</title>
<style>
  body { font-family: 'Noto Sans SC', system-ui, sans-serif; background: #0a1628; color: #e8ecf1; margin: 0; padding: 32px; }
  h1 { font-size: 20px; margin: 0 0 4px; }
  .meta { color: #556677; font-size: 12px; margin-bottom: 24px; }
  .cards { display: flex; gap: 12px; margin-bottom: 24px; }
  .card { background: #121e33; border: 1px solid #1e3454; border-radius: 8px; padding: 12px 20px; }
  .card .v { font-size: 20px; font-weight: 700; color: #00a8ff; }
  .card .l { font-size: 11px; color: #556677; }
  .msg { margin-bottom: 12px; padding: 10px 14px; border-radius: 8px; background: #121e33; border: 1px solid #1e3454; }
  .msg.user { border-left: 3px solid #00a8ff; }
  .msg.assistant { border-left: 3px solid #ff9500; }
  .msg.system { background: transparent; border: none; color: #8899aa; font-size: 12px; }
  .tag { font-size: 11px; font-weight: 700; color: #8899aa; margin-right: 8px; }
  .time { font-size: 11px; color: #556677; }
  .content { margin-top: 6px; line-height: 1.6; white-space: pre-wrap; word-break: break-word; }
</style>
</head>
<body>
  <h1>${esc(session.title)}</h1>
  <div class="meta">创建: ${esc(new Date(session.createdAt).toLocaleString('zh-CN'))} · 更新: ${esc(new Date(session.updatedAt).toLocaleString('zh-CN'))}</div>
  <div class="cards">
    <div class="card"><div class="v">${stats.messages}</div><div class="l">消息总数</div></div>
    <div class="card"><div class="v">${stats.user}</div><div class="l">用户消息</div></div>
    <div class="card"><div class="v">${Math.round(stats.duration / 1000)}s</div><div class="l">会话时长</div></div>
  </div>
  ${rows}
</body>
</html>`
}

/** 生成报告并触发浏览器下载 */
export function generateReport(session: Session): void {
  const html = buildHtml(session)
  const blob = new Blob([html], { type: 'text/html;charset=utf-8' })
  const url = URL.createObjectURL(blob)
  const a = document.createElement('a')
  const safeTitle = session.title.replace(/[\\/:*?"<>|]/g, '_').slice(0, 30)
  a.href = url
  a.download = `会话报告_${safeTitle}_${new Date().toISOString().slice(0, 10)}.html`
  document.body.appendChild(a)
  a.click()
  a.remove()
  URL.revokeObjectURL(url)
}
